test/src/test/java/org/corfudb/runtime/view/LogUnitSealTest.java

```diff
@@ -0,0 +1,105 @@
+        addServer(SERVERS.PORT_0);
+
+        Layout l = new TestLayoutBuilder()
+                .setEpoch(epoch)
+                .addLayoutServer(SERVERS.PORT_0)
+                .addSequencer(SERVERS.PORT_0)
+                .buildSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_0)
+                .addToSegment()
+                .addToLayout()
+                .build();
+
+        bootstrapAllServers(l);
+        CorfuRuntime corfuRuntime = getRuntime(l).connect();
+        RuntimeLayout runtimeLayout = corfuRuntime.getLayoutView().getRuntimeLayout(l);
+
+        getManagementServer(SERVERS.PORT_0).shutdown();
+
+        return runtimeLayout;
+    }
+
+    /**
+     * Test that log unit server seals correctly.
+     * Check operations in the queue that come after the seal operation, but stamped with
+     * an older epoch, are completed exceptionally with a WrongEpochException.
+     */
+    @Test
+    public void checkOperationWithOldEpochFailedAfterSeal() {
+        long epoch = 1L;
+        RuntimeLayout runtimeLayout = getRuntimeLayout(epoch);
+        Layout layout = runtimeLayout.getLayout();
+        LogUnitClient client = runtimeLayout.getLogUnitClient(SERVERS.ENDPOINT_0);
+
+        List<CompletableFuture<Boolean>> successOpFutures1 = new ArrayList<>();
+        List<CompletableFuture<Boolean>> successOpFutures2 = new ArrayList<>();
+        List<CompletableFuture<Boolean>> failedOpFutures = new ArrayList<>();
+
+        final int numOp = 3;
+        long address = 0L;
+
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            successOpFutures1.add(client.write(ld));
+        }
+
+        // Seal server
+        layout.setEpoch(layout.getEpoch() + 1);
+        runtimeLayout.sealMinServerSet();
+
+        // Intentionally reset router epoch to bypass epoch check on router.
+        getLayoutServer(SERVERS.PORT_0).getServerContext().getServerRouter().setServerEpoch(epoch);
+
+        // Still using the old client to send messages stamped with old epoch
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            failedOpFutures.add(client.write(ld));
+        }
+
+        // Set the router epoch to new epoch and get the new client
+        getLayoutServer(SERVERS.PORT_0).getServerContext().getServerRouter().setServerEpoch(layout.getEpoch());
+        runtimeLayout = new RuntimeLayout(layout, getRuntime(layout));
+        client = runtimeLayout.getLogUnitClient(SERVERS.ENDPOINT_0);
+
+        // Using the new client to send messages stamped with epoch 2
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            successOpFutures2.add(client.write(ld));
+        }
+
+        // Before sealing, all operations should complete normally
+        successOpFutures1.forEach(f -> assertThatCode(f::join).doesNotThrowAnyException());
```

Comment:
Does this actually wait on the futures to complete? 
If not, then there could be a race condition, where this check is executed before the future is completed.

Reply:
Yes, assertThatCode() will wait on future to complete, I also did an experiment to verify it.